File: day1/perl/day1.pl
#!/usr/bin/perl
$file = "../input";

print "--- Part One ---\n";
open $input_fh, '<', $file or die "Could not open $file: $!\n";

$count = 0;
$prev_value = <$input_fh>;
$value = <$input_fh>;
chop($prev_value);
chop($value);
while ($line = <$input_fh>) {
	if ($prev_value < $value) {
		$count++;
	}
	$prev_value = $value;
	$value = $line;
	chop($value);
}
if ($prev_value < $value) {
	$count++;
}
print "Result: ".$count." 'increased'\n";
close $input_fh or die "Failed to close $file: $!\n";
print "--- Part Two ---\n";

sub sum_array {
	my @array = @{$_[0]};
	my $sum = 0;
	foreach my $i (@array) {
		$sum += $i
	}
	return $sum;
}

open $input_fh, '<', $file or die "Could not open $file: $!\n";

$count = 0;
@window1 = (0, 0, 0);
$window1[0] = <$input_fh>;
chop($window1[0]);
$window1[1] = <$input_fh>;
chop($window1[1]);
$window1[2] = <$input_fh>;
chop($window1[2]);
@window2 = ($window1[1], $window1[2], 0);
$window2[2]=<$input_fh>;
chop($window2[2]);
while ($line = <$input_fh>) {
#	print "[".$window1[0].", ".$window1[1].", ".$window1[2]."] - ";
#	print "[".$window2[0].", ".$window2[1].", ".$window2[2]."]\n";
#	print sum_array(\@window1)." vs ".sum_array(\@window2)."\n";
	if (sum_array(\@window1) < sum_array(\@window2)) {
		$count++;
	}
	$window1[0] = $window1[1];
	$window1[1] = $window1[2];
	$window1[2] = $window2[2];
	$window2[0] = $window2[1];
	$window2[1] = $window2[2];
	$window2[2] = $line;
	chop($window2[2]);
}
if (sum_array(\@window1) < sum_array(\@window2)) {
	$count++;
}

print "Result: ".$count." 'increased'\n";
close $input_fh or die "Failed to close $file: $!\n";

File: day1/go/day1.go
package main

import (
	"bufio"
	"log"
	"fmt"
	"os"
	"strconv"
)

func part1() {
	fmt.Println("--- Part One ---");
	file, err := os.Open("../input")
	if err != nil {
		log.Fatal(err)
		os.Exit(1)
	}
	defer file.Close()
	count := 0
	scanner := bufio.NewScanner(file)
	scanner.Scan()
	if err := scanner.Err(); err != nil {
		log.Fatal(err)
		os.Exit(1)
	}
	prev_value, err := strconv.Atoi(scanner.Text())
	scanner.Scan()
	value, err := strconv.Atoi(scanner.Text())
	for scanner.Scan() {
		if prev_value < value {
			count++
		}
		prev_value = value
		value, err = strconv.Atoi(scanner.Text())
	}
	if prev_value < value {
		count++
	}
	fmt.Printf("Result: %d 'increased'\n", count);
}

func sum_array(array []int) int {
	sum := 0

	for i := range array {
		sum += array[i]
	}
	return sum
}

func part2() {
	fmt.Println("--- Part One ---");
	file, err := os.Open("../input")
	if err != nil {
		log.Fatal(err)
		os.Exit(1)
	}
	defer file.Close()
	count := 0
	scanner := bufio.NewScanner(file)
	var window1[3]int
	var window2[3]int
	for i := range window1 {
		scanner.Scan()
		if err := scanner.Err(); err != nil {
			log.Fatal(err)
			os.Exit(1)
		}
		window1[i], err = strconv.Atoi(scanner.Text())
	}
	for i := range window2[:len(window2) - 1]{
		window2[i] = window1[i + 1]
	}
	scanner.Scan()
	window2[2], err = strconv.Atoi(scanner.Text())
	for scanner.Scan() {
		if sum_array(window1[:]) < sum_array(window2[:]) {
			count++
		}
		for i := range window1 {
			window1[i] = window2[i]
		}
		for i := range window2[:len(window2) - 1]{
			window2[i] = window2[i + 1]
		}
		window2[len(window2) - 1], err = strconv.Atoi(scanner.Text())
	}
	if sum_array(window1[:]) < sum_array(window2[:]) {
		count++
	}
	fmt.Printf("Result: %d 'increased'\n", count);

}

func main() {
	part1();
	part2();
}
